x-pack/legacy/plugins/siem/public/containers/source/index.tsx

```diff
@@ -110,3 +113,56 @@ WithSource.displayName = 'WithSource';
 
 export const indicesExistOrDataTemporarilyUnavailable = (indicesExist: boolean | undefined) =>
   indicesExist || isUndefined(indicesExist);
+
+export const useWithSource = (sourceId: string, indices: string[]) => {
+  const [loading, updateLoading] = useState(false);
+  const [indicesExist, setIndicesExist] = useState<boolean | undefined | null>(undefined);
+  const [browserFields, setBrowserFields] = useState<BrowserFields | null>(null);
+  const [indexPattern, setIndexPattern] = useState<StaticIndexPattern | null>(null);
+  const [errorMessage, updateErrorMessage] = useState<string | null>(null);
+
+  const apolloClient = useApolloClient();
+  async function fetchSource(signal: AbortSignal) {
+    updateLoading(true);
+    if (apolloClient) {
+      apolloClient
+        .query<SourceQuery.Query, SourceQuery.Variables>({
+          query: sourceQuery,
+          fetchPolicy: 'cache-first',
+          variables: {
+            sourceId,
+            defaultIndex: indices,
+          },
+          context: {
+            fetchOptions: {
+              signal,
+            },
+          },
+        })
+        .then(
+          result => {
+            updateLoading(false);
+            updateErrorMessage(null);
+            setIndicesExist(get('data.source.status.indicesExist', result));
+            setBrowserFields(getBrowserFields(get('data.source.status.indexFields', result)));
+            setIndexPattern(
+              getIndexFields(indices.join(), get('data.source.status.indexFields', result))
```

Comment:
Optional, no changes requested: You can use the fancy new `?.` if you want here I think from TypeScript.